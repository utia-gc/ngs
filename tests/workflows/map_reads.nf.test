nextflow_workflow {

    name "Test Workflow MAP_READS"
    script "workflows/map_reads.nf"
    workflow "MAP_READS"

    test("MAP_READS maps reads to a reference genome and outputs sorted BAMs -- bwa-mem2 + SE reads.") {
        // instantiate helper objects
        def readsLane1 = new ReadsSELane1()
        def refs       = new RefsEnsembl()

        when {
            params {
                publishDirReports = "${outputDir}/reports"
                publishDirData    = "${outputDir}/data"
            }
            workflow {
                """
                input[0] = [
                    [
                        sampleName:   "${readsLane1.getSampleName()}",
                        sampleNumber: "${readsLane1.getSampleNumber()}",
                        readType:     "${readsLane1.getReadType()}"
                    ],
                    [
                        file("${readsLane1.getR1()}"),
                    ]
                ]
                input[1] = file("${refs.getGenome()}")
                input[2] = "bwa-mem2"
                """
            }
        }

        then {
            // test workflow status
            assert workflow.success

            // test alignments
            with(workflow.out.alignments.get(0)) {
                assert size() == 3
                // test alignments file
                assert get(1) ==~ /^.*\/${readsLane1.getSampleName()}_sorted\.bam/
                // test indexed alignments file
                assert get(2) ==~ /^.*\/${readsLane1.getSampleName()}_sorted\.bam\.bai/
            }
            // test map log
            // with(workflow.out.map_log) {
            //     assert isEmpty()
            // }

            // test reproducibility
            assert snapshot(workflow.out).match()
        }

    }

    test("MAP_READS maps reads to a reference genome and outputs sorted BAMs -- bwa-mem2 + PE reads.") {
        // instantiate helper objects
        def readsLane1 = new ReadsPELane1()
        def refs       = new RefsEnsembl()

        when {
            params {
                publishDirReports = "${outputDir}/reports"
                publishDirData    = "${outputDir}/data"
            }
            workflow {
                """
                input[0] = [
                    [
                        sampleName:   "${readsLane1.getSampleName()}",
                        sampleNumber: "${readsLane1.getSampleNumber()}",
                        readType:     "${readsLane1.getReadType()}"
                    ],
                    [
                        file("${readsLane1.getR1()}"),
                    ]
                ]
                input[1] = file("${refs.getGenome()}")
                input[2] = "bwa-mem2"
                """
            }
        }

        then {
            // test workflow status
            assert workflow.success

            // test alignments
            with(workflow.out.alignments.get(0)) {
                assert size() == 3
                // test alignments file
                assert get(1) ==~ /^.*\/${readsLane1.getSampleName()}_sorted\.bam/
                // test indexed alignments file
                assert get(2) ==~ /^.*\/${readsLane1.getSampleName()}_sorted\.bam\.bai/
            }
            // test map log
            // with(workflow.out.map_log) {
            //     assert isEmpty()
            // }

            // test reproducibility
            assert snapshot(workflow.out).match()
        }

    }

}
