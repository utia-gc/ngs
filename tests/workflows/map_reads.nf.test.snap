{
    "MAP_READS maps reads to a reference genome and outputs sorted BAMs -- bwa-mem2 + SE reads.": {
        "content": [
            {
                "0": [
                    [
                        {
                            "sampleName": "SRR1066657",
                            "sampleNumber": "3",
                            "readType": "single"
                        },
                        "SRR1066657_sorted.bam:md5,7714ba0359a7ee81188527413e95a085",
                        "SRR1066657_sorted.bam.bai:md5,b84e9453531a33ca4c05ec65637a21de"
                    ]
                ],
                "alignments": [
                    [
                        {
                            "sampleName": "SRR1066657",
                            "sampleNumber": "3",
                            "readType": "single"
                        },
                        "SRR1066657_sorted.bam:md5,7714ba0359a7ee81188527413e95a085",
                        "SRR1066657_sorted.bam.bai:md5,b84e9453531a33ca4c05ec65637a21de"
                    ]
                ]
            }
        ],
        "timestamp": "2023-09-25T19:03:47+0000"
    },
    "MAP_READS maps reads to a reference genome and outputs sorted BAMs -- bwa-mem2 + PE reads.": {
        "content": [
            {
                "0": [
                    [
                        {
                            "sampleName": "SRR6924569",
                            "sampleNumber": "1",
                            "readType": "paired"
                        },
                        "SRR6924569_sorted.bam:md5,17906e3b45c20534269939c7cdae044c",
                        "SRR6924569_sorted.bam.bai:md5,3fffc82ecc2424f07ef81e35f51028f6"
                    ]
                ],
                "alignments": [
                    [
                        {
                            "sampleName": "SRR6924569",
                            "sampleNumber": "1",
                            "readType": "paired"
                        },
                        "SRR6924569_sorted.bam:md5,17906e3b45c20534269939c7cdae044c",
                        "SRR6924569_sorted.bam.bai:md5,3fffc82ecc2424f07ef81e35f51028f6"
                    ]
                ]
            }
        ],
        "timestamp": "2023-09-25T19:03:47+0000"
    }
}